### PR TITLE
issue-1517 - Support for variables in JSON configuration files like 'fhir-server-config.json'

### PR DESCRIPTION
**Pull request by holger-stenzhorn**
Use Apache Commons Text’s StringSubstitutor in ConfigurationService’s loadConfiguration method to replace variables in JSON configuration files with the values of the respective envirnoment variables.

### CODE REVIEW

File: fhir-config/pom.xml
Comment (prb112) [inline fhir-config/pom.xml:43]:
```suggestion
```
duplicate declaration as on line 44

File: fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java
Comment (prb112) [inline fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java:9]: note, we should probably change the copyright year to 2020

File: fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java
Comment (holger-stenzhorn) [reply fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java:62]: The use of such a prefix is of course possible and might be helpful to distinct environment variables targeted for the FHIR server from the rest - yet I am just not convinced whether this should be mandatory. Rather I believe that such hard-coded prefix could possibly restrict the user from fully following and complying to his/her own, local naming conventions for variables. Also OpenLiberty itself does not stipulate any prefix either for (environment) variables used in its XML configuration files, if I see that correctly. What do you think?

Comment (prb112) [reply fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java]: I think it's best to prefix.  While it constrains the environment, it doesn't permit needless introspection into the environment variables that may be beyond the scope of the application.  

Maybe it's best to have a second reviewer comment? @lmsurpre thoughts?

Comment (holger-stenzhorn) [reply fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java]: I am sorry but I must admit that I do not understand what you mean with "permit needless introspection into the environment variables that may be beyond the scope of the application". The substitution of the variables within the JSON configuration files happens directly "on-the-fly" when loading those files within `ConfigurationService.loadConfiguration(InputStream is)` and hence the actual application, i.e. the FHIR server, does not see the variables themselves but only their (final) values and thus does not have to perform any further introspection.

Comment (prb112) [reply fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java]: Right, I understand that, however this code also is permitted to run in a non-docker environment.  An attacker could conceivably gain access to more info than required, if not prefixed.

Comment (holger-stenzhorn) [reply fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java]: I am sorry again for my ignorance but I am do not get the point how a prefix `FHIR.` could help to have an attacker gain access to less information. For example, using `postgres.password` as plain environment variable is definitely "suboptimal" security-wise but I do not understand how `FHIR.postgres.password` would ameliorates things. Can you perhaps give me an example/use case with a "malicious" variable so that I can better get what you mean? That would be great!

Comment (prb112) [reply fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java]: So it's a reference to a secondary attack, where you inject a new configuration which may have access to an unexpected or unscoped environment variable. 

Also, a good lesson from CERT https://wiki.sei.cmu.edu/confluence/display/c/ENV02-C.+Beware+of+multiple+environment+variables+with+the+same+effective+name and more specific to java https://wiki.sei.cmu.edu/confluence/display/java/ENV02-J.+Do+not+trust+the+values+of+environment+variables

I like the thought of scoping it as it is an intent to use versus a side effect. 

PRO
no changes, user defines completely

CON
indirect re-use of environment variables (potential overrides)
potential conflicts with other Environment variables

Comment (holger-stenzhorn) [reply fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java]: Thanks a lot for providing me with the information and links. Yet I do not understand how the prefix could solve the linked issues: Following the first link, this only makes the case for having clearly distinct environment variable names without relying on case difference - so if you have `FHIR.some.variable` set to a certain value then a malicious attacker could also analogously change the prefix casing to "redefine" that value by simply setting e.g. `Fhir.some.variable`, right? The second link makes the case for using pre-defined Java system *properties* and does not mention the usage of prefixes at all - and so at the end of the day it boils down to simply not use freely defined environment variables at all - no matter if prefixed or not, right? So in the end the coping by introducing the prefix seems to be actually only a pure naming convention but would have no restrictive effect or similar - or am I still missing anything possibly?

Comment (prb112) [reply fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java]: I'm still leaning towards prefixing.  I'll leave it to John and Lee to make the final call.

**Review by prb112 — APPROVED** [review]
I'm OK with the change. It's not my preference to use a expansive namespace with environment variables. I prefer scoping the namespace, but understand that the author does not.  That's fine.  difference of engineering. 

If there are concerns with security scans, we'll deal with them in-one offs.

**Review by lmsurpre — APPROVED** [review]
LGTM

**Comment by lmsurpre** [comment]
Thanks @holger-stenzhorn !

**Comment by holger-stenzhorn** [comment]
Closes #1517

**Pull request merged**
